Balanced tree creator (BTC) and probabilistic tree creator (PTC2) for GP

foolnotion opened the pull request:
Hi,

I'd like to propose a new tree creator for GP. 
The idea is similar to [PTC2](https://cs.gmu.edu/~sean/papers/treecreation.pdf), in that the creator is given a target tree length and creates an expression of that length.

The algorithm works like this:
- the algorithm starts from a root node randomly sampled from the primitive set and keeps track of a horizon of expansion points (node child slots yet unfilled by the algorithm)
- the slots are filled in breadth-first fashion while keeping track of the remaining number of nodes necessary to reach the target tree length
- when a function node is added to the tree, it opens a number of new expansion points according to the function arity
- the arity of nodes sampled from the primitive set is then limited according to the difference between the target length and the remaining number of unfilled expansion points
- when the difference becomes zero, the algorithm fills the remaining expansion points with leaf nodes
- a `bias` parameter can affect the shape of the created trees (fully balanced or a bit more random). the trees by default are flat due to the breadth-first filling scheme. the idea is to allow the possibility of a leaf node to be sampled instead of a function node within the expansion loop, as long as this does not prevent the algorithm from reaching the target length.
- complexity is linear O(n) where n is the length

Using this tree creator one can generate trees obeying any length distribution:
[balanced_normal.pdf](https://github.com/DEAP/deap/files/4674388/balanced_normal.pdf)
[balanced_uniform.pdf](https://github.com/DEAP/deap/files/4674389/balanced_uniform.pdf)

I am not a python expert so please feel free to review and request changes/improvements to the code. Two function signatures are provided:
- one accepting a target length
- one accepting a distribution along with distribution args, from which to sample the length
- a max depth limit is not included because the trees will be flat anyway, but it would be trivial to add if you think it's useful

The algorithm was originally implemented in C++ [here](https://github.com/foolnotion/operon/blob/master/include/operon/operators/creator/balanced.hpp). It is described in a paper accepted to be published in GECCO EvoSoft 2020 (I'll upload an arxiv preprint soon). 

**EDIT**: I've also implemented the PTC2 algorithm described in Luke - Two Fast Tree Creation Algorithms for Genetic Programming (https://ieeexplore.ieee.org/document/873237), a pdf version can also be found [here](https://cs.gmu.edu/~sean/papers/treecreation.pdf), with two improvements:
- no maximum depth needed, always produces the desired length
- `bias` parameter can control tree shape 

Best,
Bogdan

foolnotion commented:
Test code:

```python
import numpy as np
import sys, uuid
from deap import gp
import seaborn as sns
import matplotlib.pyplot as plt

cols = 10 # 10 variables
pset = gp.PrimitiveSet("MAIN", cols)
pset.addPrimitive(np.add, 2, name="vadd")
pset.addPrimitive(np.subtract, 2, name="vsub")
pset.addPrimitive(np.multiply, 2, name="vmul")
pset.addPrimitive(np.divide, 2, name="vdiv")
pset.addPrimitive(np.negative, 1, name="vneg")
pset.addPrimitive(np.cos, 1, name="vcos")
pset.addPrimitive(np.sin, 1, name="vsin")
pset.addPrimitive(np.exp, 1, name="vexp")
pset.addPrimitive(np.log, 1, name="vlog")
pset.addEphemeralConstant("rand101_" + str(uuid.uuid4()), lambda: np.random.uniform(-1.0, 1.0)) #may be unable to pickle...

max_length = 100 

#lengths = np.random.random_integers(1, max_length, 10000)
lengths = [ 1 if l < 1 else int(l) for l in np.random.normal(50, 12, 10000) ]
trees = [ gp.genBalanced(pset, l, bias_=1) for l in lengths ]

print("Average length: ", np.mean(lengths))

fig, axes = plt.subplots(2, 1, figsize=(10, 10))
labels = np.arange(0, max_length, 10)

ax = axes[0]
sns.countplot(lengths, ax=ax)
ax.set_title('Target lengths')

ax = axes[1]
tree_lengths = [ len(t) for t in trees ]
sns.countplot(tree_lengths, ax=ax)
ax.set_title('Actual tree lengths')

for ax in axes:
    ax.set_xticks(np.arange(0, max_length, 10))
    labels[0] = 1
    ax.set_xticklabels(labels)
    ax.set_xlabel('Tree length')
plt.savefig('balanced_normal.pdf')
```

fmder commented on deap/gp.py:
```diff
@@ -514,6 +515,258 @@ def compileADF(expr, psets):
+    return genBalanced(pset, length_, bias_, type_)
+
+
+def genBalanced(pset, length_, bias_=0, type_=None):
```
Isn't this function hiding the ``genBalanced`` of line 540?

fmder commented on deap/gp.py:
```diff
@@ -514,6 +515,258 @@ def compileADF(expr, psets):
+    return genProb(pset, length_, type_)
+
+
+def genProb(pset, length_, bias_=0, type_=None):
```
Same as last comment.

fmder commented on deap/gp.py:
```diff
@@ -514,6 +515,258 @@ def compileADF(expr, psets):
 ######################################
 # GP Program generation functions    #
 ######################################
+def SampleChild(pset, minArity_, maxArity_, type_):
```
This function should be camelCase. (But we are aiming at snake_case in 2.0)

fmder commented:
Thanks for the pull request! Great work. Could you add a unit test that uses this function and asserts the result is as expected?

foolnotion commented:
> Thanks for the pull request! Great work. Could you add a unit test that uses this function and asserts the result is as expected?

Thanks for the feedback, I renamed the helper functions to `generateBalanced` and `generateProb` and I changed to camelCase. The methods already include an assert at the end to ensure the target length is always achieved. 
Could you point me to the right file for adding unit tests? I found `tests/test_creator.py` but I'm not sure what that file does.

fmder commented:
You can add the file ``tests/gp_generation.py`` or something. As for the assert, is their any condition that will make it trigger? Otherwise if it is tested in a unit test it could be removed all along.

foolnotion commented:
Sorry it took me so long, I've made some improvements to the tree generators and implemented unit tests in `deap/tests/test_gp_generation.pi`.

These generators may not deliver the actual length target when the primitive set does not allow it (e.g., it is impossible to get an even length without unary functions). In this case, instead of throwing an error I've decided to push the length target towards the nearest achievable value.

This seems to work well enough and the user should be careful to provide an appropriate primitive set. For example, lacking unary functions the tree generator will only generate odd tree lengths:

![image](https://user-images.githubusercontent.com/844222/90193811-0677d080-ddc6-11ea-8d3c-97f383812cc0.png)
